muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/dao/CommunicatorMessageDAO.java

```diff
@@ -213,75 +213,68 @@ public CommunicatorMessageId findNewerThreadId(UserEntity userEntity, Communicat
-    );
+    EntityManager entityManager = getEntityManager();
+    
+    // Trashed sent messages
+    
+    CriteriaBuilder messageCriteriaBuilder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<CommunicatorMessage> messageCriteria = messageCriteriaBuilder.createQuery(CommunicatorMessage.class);
+    Root<CommunicatorMessage> messageRoot = messageCriteria.from(CommunicatorMessage.class);
+    messageCriteria.select(messageRoot);
+    List<Predicate> messagePredicates = new ArrayList<Predicate>();
+    messagePredicates.add(messageCriteriaBuilder.equal(messageRoot.get(CommunicatorMessage_.sender), user.getId()));
+    messagePredicates.add(messageCriteriaBuilder.equal(messageRoot.get(CommunicatorMessage_.trashedBySender), Boolean.TRUE));
+    messagePredicates.add(messageCriteriaBuilder.equal(messageRoot.get(CommunicatorMessage_.archivedBySender), Boolean.FALSE));
+    messageCriteria.where(messageCriteriaBuilder.and(messagePredicates.toArray(new Predicate[0])));
+    List<CommunicatorMessage> trashedMessages = entityManager.createQuery(messageCriteria).getResultList();
+    
+    // Trashed received messages
+    
+    CriteriaBuilder recipientCriteriaBuilder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<CommunicatorMessage> recipientCriteria = recipientCriteriaBuilder.createQuery(CommunicatorMessage.class);
+    Root<CommunicatorMessageRecipient> recipientRoot = recipientCriteria.from(CommunicatorMessageRecipient.class);
+    recipientCriteria.select(recipientRoot.get(CommunicatorMessageRecipient_.communicatorMessage));
+    List<Predicate> recipientPredicates = new ArrayList<Predicate>();
+    recipientPredicates.add(recipientCriteriaBuilder.equal(recipientRoot.get(CommunicatorMessageRecipient_.recipient), user.getId()));
+    recipientPredicates.add(recipientCriteriaBuilder.equal(recipientRoot.get(CommunicatorMessageRecipient_.archivedByReceiver), Boolean.FALSE));
+    recipientPredicates.add(recipientCriteriaBuilder.equal(recipientRoot.get(CommunicatorMessageRecipient_.trashedByReceiver), Boolean.TRUE));
+    recipientCriteria.where(recipientCriteriaBuilder.and(recipientPredicates.toArray(new Predicate[0])));
+    List<CommunicatorMessage> trashedReceivedMessages = entityManager.createQuery(recipientCriteria).getResultList();
+    trashedMessages.addAll(trashedReceivedMessages);
+    
+    // Sort by thread, latest message in thread first
+    
+    Comparator<CommunicatorMessage> comparator = new Comparator<CommunicatorMessage>() {
+      @Override
+      public int compare(CommunicatorMessage o1, final CommunicatorMessage o2) {
+        int result = o1.getCommunicatorMessageId().getId().compareTo(o2.getCommunicatorMessageId().getId());
+        return result == 0 ? o2.getCreated().compareTo(o1.getCreated()) : result;
+      }
+    };
+    trashedMessages.sort(comparator);
+    
+    // Filter to only the latest message in each thread
+    
+    if (!trashedMessages.isEmpty()) {
+      int i = 0;
+      while (true) {
```

Comment:
maybe `while (i < trashedMessages.size())` and remove the if with break inside the loop, at least looks safer than the scary while(true) :)